### PR TITLE
ENH: Add more initialization methods for HessianUpdateStrategy

### PR DESCRIPTION
**Pull request by jonas-eschle**
#### What does this implement/fix?
HessianUpdateStrategy takes an `init_scale` which is currently only allowed to be a scalar. However, it can be that more information about the initial hessian or inv hessian is available (for different reasons). Therefore, it should be able to incorporate this knowledge.

The change removes the float conversion and just multiplies the initial matrix by scale (whatever it is, Pythonic duck typing), except if it has the same shape as the initial matrix, in which case the initial matrix is replaced by the `init_scale`

### CODE REVIEW

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (tylerjereddy) [inline scipy/optimize/tests/test_hessian_update_strategy.py:55]: has the random seed been pinned somewhere else in the file?

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (tylerjereddy) [inline scipy/optimize/tests/test_hessian_update_strategy.py:58]: Any reason not to use `np.arange()` directly here?

**Review by tylerjereddy — COMMENTED** [review]
Are there any backwards compatibility concerns here? (I'm not an expert)

The volume of whitespace changes in the test file is a bit distracting relative to the genuine code changes there, but fair enough I suppose.

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (mayou36) [reply scipy/optimize/tests/test_hessian_update_strategy.py]: Yep, oversight, changed. Thanks!

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (mayou36) [reply scipy/optimize/tests/test_hessian_update_strategy.py]: No, I created a static matrix now, so it ain't a problem

**Comment by mayou36** [comment]
> Are there any backwards compatibility concerns here? (I'm not an expert)

Basically no, but one potential: I've removed the conversion to `float`. I don't think this is needed for any reason: what can convert to a float can usually multiply with a float. In fact, it's rather the opposite: things may can multiply with a float but cannot be converted to one.

So the case where someone has an object that _can be converted to a float_ but _cannot be multiplied with one_ will break. However, this is intentional: It's more pythonic and aligned with the expectations to have the init_scale multiplied by the matrix and _not_ first converted.

So I don't assume that this breaks any existing code.

> The volume of whitespace changes in the test file is a bit distracting relative to the genuine code changes there, but fair enough I suppose.

Indeed, that was my auto-format. It is now PEP8 conform, which I think this is not a bad thing. But I can revert it if requested

**Comment by ilayn** [comment]
The test is still too wide though

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [inline scipy/optimize/_hessian_update_strategy.py:267]: This means I can have vectors that can be broadcasted or elementwise-multiplied square matrices? We need some explanation here like anything that can be converted to a float scalar or an nxn matrix or whatever the size can be. The edge cases are more important than the happy-flow to prevent issues later

**Comment by mayou36** [comment]
> The test is still too wide though

Can you elaborate? Too wide in terms of scope? I think that the first step is basically the initialization, that's why it is there. Would you prefer to have a separate test function?

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (ilayn) [inline scipy/optimize/tests/test_hessian_update_strategy.py:67]: I meant in terms of character length here even though we don't track PEP8 in test files there is no need to have extra violations.

File: scipy/optimize/_hessian_update_strategy.py
Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:267]: Yes exactly, that's the idea, but I'll expand the text

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [inline scipy/optimize/_hessian_update_strategy.py:193]: if for some reason complex numbers with zero imaginary component comes they will upcast everything

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (mayou36) [reply scipy/optimize/tests/test_hessian_update_strategy.py]: Ah sure, I've adjusted that

File: scipy/optimize/_hessian_update_strategy.py
Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:193]: Fair point, and numpy is quite inconsistent here I figured:
```
np.array(complex(42), dtype=np.float64)  # raises TypeError
np.array(complex_array, dtype=np.float64)  # casts silently
```
So I've added an explicit check to have the same behavior as before in case of complex objects: a TypeError is raised

Comment (ilayn) [reply scipy/optimize/_hessian_update_strategy.py:193]: I would check if it is an ndarray, its `.size == 1` and if it is one push it to `float()` otherwise continue with the matrix case.

Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:193]: > I would check if it is an ndarray, its `.size == 1` and if it is one push it to `float()` otherwise continue with the matrix case.

Why would you do that? To me this seems just like extra code that won't change any of the logic. Or are you worried about implications of removing the `float` cast?

Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:193]: But sure, can be done, added it

**Comment by mayou36** [comment]
Alright, I guess this PR is ready from my side

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [reply scipy/optimize/_hessian_update_strategy.py:193]: No we don't want to stop the complex data leading to errors. That's backwards incompatible, what you are adding shouldn't touch the old behavior. Test for an instance of numpy array then if it is a scalar (from testing size ==1) it passes through float otherwise it is treated as the full hessian and continues.

Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:193]: Yes, I've added that, basically simply testing if it is size 1 -> old behavior else new

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [inline scipy/optimize/_hessian_update_strategy.py:201]: I meant this part. This wasn't here and it will block the scalar route. We only need `if isinstance(scale, ndarray)` route before the `float()`, this here introduces new behavior. Previously it was all passing through float so only new options should be redirected anything else would still go through float.

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [reply scipy/optimize/_hessian_update_strategy.py:193]: I've clarified below

File: scipy/optimize/_hessian_update_strategy.py
Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:201]: Sidenote: it may be correct to do an elif instead of an if, but since a Python float surely never is a complex object, this will never trigger the second if so it won't change the discussion

For the rest, I rather disagree:

> this here introduces new behavior

Yes, that is the _very goal_ of the PR! Adding more possibilities _will_ change the behavior. Some input in the past will now trigger different behavior. E.g. an np.array threw an error, now it won't and can be used.

However, that is fine. It is never guaranteed or documented (!) that an np.array throws an error. It was illegal and the consequences have never been defined.

However, we can catch the _reasonable_ cases such as if a complex value is given, one _may_ relies on the fact that an error is thrown. Giving in an array and getting a TypeError I would not count to it though.

_why not simply `isinstance`_: This again puts very tight constraints on the objects. What about other array-like objects? Tensors? No need to restrict us to arrays. Anything that has a shape and can handle a multiplication should be allowed.

So my take is: this will change the behavior. However, to restore the "reasonable" legacy behavior, we can say that the behavior of the function for size-1 objects (which is a large definition) remains the same. For anything else, it will change.

What do you think? Or rather, which input will now change the behavior in the past _and is worth to be conserved_?

Comment (ilayn) [reply scipy/optimize/_hessian_update_strategy.py:201]: Unfortunately there are many packages depending on us and we are tied up to keeping everything intact unless a very compelling need arises. New features are fine but old behavior needs to stay exactly as it is. If I passed 3+0.j that was passing now it is not. That's a no go regardless whether this is good or not.

Anything that was undocumented was kind of hanging in the balance but here you are clearly blocking that is not really needed to be blocked. That is the problem. There is no benefit gained from this block. My earlier comment was about since you removed the float() initially everything was passing in and numpy would upcast everything in the case of an accidental complex number. These are different issues. If you put down one guard you have to replace it somewhere else.

Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:201]: > Unfortunately there are many packages depending on us and we are tied up to keeping everything intact unless a very compelling need arises. New features are fine but old behavior needs to stay exactly as it is. 

Of course, and I fully agree. But there is still the question on what is "reasonable": there are version numbers on packages for a certain reason; this does not mean you should break features all the time but really weird behavior, well yes. That may just changes. Or there should be a difference between feature (stays the same, no question), behavior (can be reasonable to keep the same) and "undefined behavior" basically. The latter is what I would say is by passing an array in now.

> If I passed 3+0.j that was passing now it is not. That's a no go regardless whether this is good or not.

No, it was never passing. And it will not fail because of the `np.iscomplexobj` but because of the `float` that was already there before.

To be precise, it trows `TypeError: can't convert complex to float`

It always an error. Having the check for complex and throwing an error for larger objects just _aligns_ the new additional feature behavior with what was there.

Comment (mayou36) [reply scipy/optimize/_hessian_update_strategy.py:201]: So to be clear: whenever a `np.size == 1` object is passed, the behavior is *exactly* the same.

This includes a whole array of things, also complex numbers.

(_okay: under the assumption that `matrix * float` is the same as `matrix * np.array(float)`, which is the case_)

P.S: NOT under the assumption that the float is a float, that we know

**Comment by mayou36** [comment]
Hi, is there anything new on this? I think the PR is fine from my side, also backwards compatible. Can this get merged?

**Comment by mayou36** [comment]
Hi reviewers, @ilayn , @tylerjereddy , can we get this merged maybe?

**Comment by mdhaber** [comment]
@jonas-eschle are you still interested in this? It looks like the others got busy. If you are still interested in pushing this through, please remove the whitespace changes, and help me catch up - is there a reference you can point me to, or is this based on general knowledge? Do the changes just seem like a good idea, or are there examples of how this improves the performance in some way? Thanks!

**Comment by jonas-eschle** [comment]
Hi @mdhaber, apologies for this, I must have missed it! I am still interested to get this in. Can you re-open?

There is not a lot of references, I think. The problem is simple: The Hessian update has to start somewhere. This is currently a scalar or a diagonal with a scalar.
But it's very well possible that more is already known, not just diag(scalar). Maybe the whole Hessian matrix is already known (for example if you first minimize approximately with one algorithm or you restart a fit with a very similar setup).

This allows to give the whole matrix if known, this was not possible before because of a "float" cast (which should never have been there in the first place, tbh).
Also, it is more verbose if you give a complex number (which doesn't work).

The whitespace are "free fixes", [scipy adheres to PEP8](https://docs.scipy.org/doc/scipy/dev/contributor/pep8.html) AFAIU, which [recommends spaces where I put theme](https://peps.python.org/pep-0008/#other-recommendations).

**Comment by mdhaber** [comment]
Sure; reopened. There is one conflict that should be easy for you to resolve.

> The whitespace are "free fixes", scipy adheres to PEP8 AFAIU, which recommends spaces where I put theme.

PEP8 does not recommend all of these adjustments. It explicitly leaves many of them to personal taste, and uniformly adding whitespace around all operators - even those with high priority (e.g. exponentiatiation) - may be against the recommendation.

image

In any case:
- Whitespace changes increase the scope of review. If you want the PR to be reviewed efficiently, it's better to limit what is asked of the reviewer to the essentials (i.e. adding initialization methods, in this case).
- Changing whitespace adds churn to the history. When looking at the blame, it's better to see when the line was actually written; who changed the whitespace when is rarely of interest.

So please do remove the whitespace adjustments.

>  The problem is simple: The Hessian update has to start somewhere. This is currently a scalar or a diagonal with a scalar.

Got it. One the merge conflict is resolved and the whitespace changes are made, I may be able to help further, or I might ping someone else.

**Comment by jonas-eschle** [comment]
Great, thanks a lot for the speedy review, I've removed the whitespaces and rebased the commit.

> PEP8 does not recommend all of these adjustments. It explicitly leaves many of them to personal taste

Yes, true, I actually meant mostly the consistency part. Sometimes a line has whitespace, sometimes a line has not. But nvm, removed it anyways.

**Comment by jonas-eschle** [comment]
@mdhaber can we merge this?

**Comment by mdhaber** [comment]
@ilayn @tylerjereddy looks like you had started to review this a few years ago. Are you able to take another look?

**Comment by jonas-eschle** [comment]
@mdhaber as there was no reply and the concept of it was already discussed, can we maybe get this merged?

**Comment by mdhaber** [comment]
I think the best I can do is to ask @ilayn @tylerjereddy to resume their reviews. This is not in my area of interest, and the time when I might have been able to get into it has passed.

**Comment by jonas-eschle** [comment]
@mdhaber realistically speaking, it would be great to have this in after all the years. For the change itself, it's trivial: The HessianUpdateStrategy approximates a Hessian (i.e. n x n) matrix. This can currently be initialized with a scalar that is then broadcast to a matrix. The PR simply allows to initialize this with a matrix. That's all there is.

I do understand scrutiny, but this is a trivial change, all is there and nobody disagreed of the reviewers (regarding the two year old posts). Any comments had been addressed. All the tests pass. New tests pass as well. It's backwards compatible. What else is there to risk?
I would suggest to give another try and find someone to review and otherwise just merge it?

**Comment by mdhaber** [comment]
You are welcome to look for reviewers by posting at https://discuss.scientific-python.org/c/contributor/scipy.

**Comment by jonas-eschle** [comment]
> You are welcome to look for reviewers by posting at [discuss.scientific-python.org/c/contributor/scipy](https://discuss.scientific-python.org/c/contributor/scipy).

Hm, doesn't seem like the right place for a small review tbh? But I've adjusted the title, it's meaningless without review, so it's clear where it's stuck

**Comment by ilayn** [comment]
I'll get to this in a few days the latest, apologies for the delay

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for your contribution and patience @jonas-eschle !

This looks mostly done. What I am missing is stricter input validation: `init_scale` must be a two dimensional symmetric matrix if it is provided as an array from what i understand. If a user provides a 3 D array or a nonsymmetric array, an error should be raised. Could you add this check and tests that match the error message?

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (dschmitz89) [inline scipy/optimize/tests/test_hessian_update_strategy.py:104]: Sorry for being not verbose enough. At least in `scipy.stats`, the preferred way to test such input validation errors is to use `pytest.raises(ErroType, match="error message")`. See for example [here](https://github.com/scipy/scipy/blob/63ca1a1ccea554cedd0687e986a12cc6d27c924a/scipy/stats/tests/test_multivariate.py#L857) for a similar case. Could you please adapt your tests to this pattern? Then I will approve from the `optimize` side.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]

File: scipy/optimize/_hessian_update_strategy.py
Comment (dschmitz89) [reply scipy/optimize/_hessian_update_strategy.py]: I have never used complex numbers in Python but I would be very surprised if any nonlinear optimizer in  `scipy.optimize` worked with them. In principle, I agree with @jonas-eschle that if anyone relied on supplying a complex number here was using a bug not a feature. Still, backwards compatibility ..

Could we perform the complex number check only for the newly added array possibility? That would avoid changing old behaviour.

File: scipy/optimize/tests/test_hessian_update_strategy.py
Comment (jonas-eschle) [reply scipy/optimize/tests/test_hessian_update_strategy.py]: Sure, I prefer that too. I've parametrized in a loop, but I can also use the pytest.mark.parametrized if preferred

File: scipy/optimize/_hessian_update_strategy.py
Comment (jonas-eschle) [reply scipy/optimize/_hessian_update_strategy.py]: We already did this here: the legacy way was to cast to a float, and that's exactly what we're doing now as well. This _will_ fail for a complex value, so the `np.iscomplex` will never be reached and only applies to the new arrays. But I just changed the code a bit anyways, just to be more specific with shapes etc, and that naturally included an `elif`

The only non-backwards compatible change is for arrays, that a) won't fail anymore (that's the feature we're adding) or b) will fail for other reasons (which is now the case as I've added the additional checks).

(just as an overall sidenote, this PR could nearly be considered a bug fix instead of a new feature: it seemed quite surprising that the init value, the value to initialize an _array_, was _cast_ to a float, and passing a numpy array as initialization _was not possible_. With all the general numpy array compatibility and some freedom for broadcasts...)

**Comment by jonas-eschle** [comment]
Hi @dschmitz89 and thanks for taking the time to review! The checks make sense to me, to keep backwards compatibility very strict, I didn't put them there first, but I agree that it should be there. Added and slightly refactored. Also extended the tests. Should now all be fine from my side

**Comment by jonas-eschle** [comment]
> Thanks for your contribution and patience @jonas-eschle !
> 
> This looks mostly done. What I am missing is stricter input validation: `init_scale` must be a two dimensional symmetric matrix if it is provided as an array from what i understand. If a user provides a 3 D array or a nonsymmetric array, an error should be raised. Could you add this check and tests that match the error message?

Thanks a lot for taking the time to go through @dschmitz89! I was avoiding explicit checks on purpose

**Comment by dschmitz89** [comment]
Not sure what is causing the doc build failure. Could you merge main into your branch?

To swing more reviewers on your side, you might add a test where the new array argument results in less function evaluations than the default `auto` for an optimisation problem.

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [inline scipy/optimize/_hessian_update_strategy.py:299]: I think we can be a bit more expressive. I think if I understand correctly, an np.array without (n x n) shape is discarded silently. Hence should not it be resulting in an error instead? 

```python
"""
        This parameter can be used to initialize the Hessian or its
        inverse. When a float is given, the relevant array is initialized
        to ``np.eye(n) * init_scale``, where ``n`` is the problem dimension.
        Alternatively, if a precisely ``(n, n)`` shaped array is given, this 
        array will be used. Otherwise an error is generated.
        Set it to 'auto' in order to use an automatic heuristic for choosing
        the initial scale. The heuristic is described in [1]_, p.143.
        The default is 'auto'.
"""
```

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [inline scipy/optimize/_hessian_update_strategy.py:208]: Better to use the copy command to make sure that the data is not still not owned

**Review by ilayn — CHANGES_REQUESTED** [review]
A few very minor comments from my side. This is good to go after the changes in my opinion

File: scipy/optimize/_hessian_update_strategy.py
Comment (jonas-eschle) [reply scipy/optimize/_hessian_update_strategy.py:299]: Yes, that's nicely formulated, I added the "symmetric" requirement.

> an np.array without (n x n) shape is discarded silently. 

An error is raised already. Either a "scalar-like" (`np.size(init) == 1`) or an array with shape (n, n), symmetric, is allowed. Everything else raises an error.

File: scipy/optimize/_hessian_update_strategy.py
Comment (jonas-eschle) [reply scipy/optimize/_hessian_update_strategy.py]: I've extended the error message a bit, but not sure what exactly could be told?

(I've used matrix as in "hessian matrix", but replaced with "hess/inv_hess" as this is also used as the type keyword)

File: scipy/optimize/_hessian_update_strategy.py
Comment (jonas-eschle) [reply scipy/optimize/_hessian_update_strategy.py]: `copy` defaults to true for `np.array` (not for `np.asarray`, but I've added it explicitly.

File: scipy/optimize/_hessian_update_strategy.py
Comment (ilayn) [reply scipy/optimize/_hessian_update_strategy.py]: My bad. What I meant was to actually tell the user what exactly is wrong and what has to be the right thing. So if we already detected the array with the wrong shape we can say so

```python
                if shape != np.shape(scale):
                    raise ValueError("If init_scale is an array, it must have the "
                                     f"dimensions of the matrix ({n}, {n}) but got {shape}.")

```

if it fits into PEP8 limit otherwise with a line break somewhere

Comment (jonas-eschle) [reply scipy/optimize/_hessian_update_strategy.py]: Hm, but that's exactly what it's doing right now I think? `shape` _is_ `({n}, {n})`. I've added that it also prints out the user provided array.

It's also tested in plait string in the tests

Thanks for reviewing and the scrutiny btw, appreciate it!

Comment (ilayn) [reply scipy/optimize/_hessian_update_strategy.py]: Yeah I missed your change while I was typing my diatribe. Thank you for the rapid changes. Much appreciated. I'll mark it as resolved.

**Review by ilayn — APPROVED** [review]

**Comment by dschmitz89** [comment]
> > Not sure what is causing the doc build failure. Could you merge main into your branch?
> 
> yes, weird, I cannot find any errors either. Merged. EDIT: solved the issue.
> 
> > To swing more reviewers on your side, you might add a test where the new array argument results in less function evaluations than the default `auto` for an optimisation problem.
> 
> I'll gladly elaborate on the usefulness of this somewhere (which is: hotstart, if you have the full matrix from a previous fit, often used for example in profiling a likelihood where you start very close to the previous minimum). But not sure a test is the best thing? This will involve many elements not directly related to it. If the goal is to tell users about it, then changelog for sure, docs maybe? I think tests are for the correctness of the code itself, not per se to provide a motivation; a hotstart is desired, would that actually need a motivation?
> 
> Please let me know, I am a bit confused about the proposal and still rather new to SciPy contribution

Allright, you made your case. :) Will leave this open for a few more days in case there are more comments, and then merge.

**Review by dschmitz89 — APPROVED** [review]

**Comment by dschmitz89** [comment]
> Great, thanks a lot everyone for taking the time and effort to review!
> 
> Two minor questions remaining regarding documenting:
> - sometimes, a "versionadded" is added, but the change is also comparably minimal. Should one be added here too, what's your preference?

As we are not adding a new argument, I would not add that label here. We are not fully consistent with this in SciPy also.

> - how does the changelog get updated, exactly? I am aware of the Wiki page, but can't edit, how will this be updated? (I didn't find any guide for that)

Due to spam edits we had to remove edit permissions on that page. A maintainer will add that for you.

**Comment by ilayn** [comment]
Version added not required for changes indeed. 

Thank you @jonas-eschle  and all. Apologies again for the delays.

**Comment by jonas-eschle** [comment]
Great, thanks a lot @ilayn and all of the others for the merge and the reviews!

**Pull request merged**
